Add combobox primitive

joshuaellis opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the tests
- Create or update the documentation at https://github.com/strapi/documentation
- Refer to the issue you are closing in the PR description: Fix #issue
- Specify if the PR is ready to be merged or work in progress (by opening a draft PR)

Please ensure you read the Contributing Guide: https://github.com/strapi/strapi/blob/master/CONTRIBUTING.md
-->

### What does it do?

* Adds a new combobox primitive to be used in refactoring of the combobox made in a composable manor

### Why is it needed?

* needed to abstract the accessibility logic of the DS combobox

### How to test it?

* automated tests 🪄

joshuaellis commented:
I need to write some documentation, but from a code perspective i think this is good.

markkaylor reviewed (commented):
Some of the "stolen" code is a bit wild. I trust your judgement but wondering if it might be worth adding unit tests?

When I test the primitive in storybook, after selecting a value I am unable to clear the input, is that the expected behavior for the primitive?

joshuaellis commented on docs/package.json:
```diff
@@ -12,6 +12,7 @@
     "@storybook/react": "^6.5.15",
     "@strapi/design-system": "^1.6.3",
     "@strapi/icons": "^1.6.3",
+    "@strapi/primitives": "^1.6.1",
```
I think this will update on the next `lerna` versioning

joshuaellis commented:
> Some of the "stolen" code is a bit wild. I trust your judgement but wondering if it might be worth adding unit tests?

Can add unit tests if you'd like?

> When I test the primitive in storybook, after selecting a value I am unable to clear the input, is that the expected behavior for the primitive?

When you say "clear" the input, do you mean with a button? or at all? because you should be able to delete the string, that might be a bug.

gu-stav commented on packages/primitives/src/components/Combobox/Combobox.tsx:
```diff
@@ -0,0 +1,764 @@
+            }}
+            onKeyDown={composeEventHandlers(contentProps.onKeyDown, (event) => {
+              // Combobox should not be navigated using tab key so we prevent it
+              if (event.key === 'Tab') event.preventDefault();
```
Since this is handled for all cases at the end I don't think you need to handle this separately?

joshuaellis replied:
It's not handled for all cases, just `['ArrowUp', 'ArrowDown', 'Home', 'End']`

gu-stav replied:
Ah missed the closing bracket. Mh - what do you think of grouping these together?

```
if (['ArrowUp', 'ArrowDown', 'Home', 'End', 'Tab'].includes(event.key)) {
  event.preventDefault();
}
```

at the end?

joshuaellis replied:
I don't really mind but I found it easier to understand:

* this key does that
* these keys do this

instead of:
* these keys do this
* all these keys prevent default though

What makes you prefer the latter?

gu-stav replied:
I personally find the latter case you've described easier to understand, but since is just a coding-style thing and your solution is in-line with all the other places let's keep it 👍🏼

joshuaellis commented:
> Yeah when you try to delete the string it just refills once the value is empty: [design-system-git-feature-listbox-strapijs.vercel.app/?path=/story/design-system-primitives-combobox--basic-usage](https://design-system-git-feature-listbox-strapijs.vercel.app/?path=/story/design-system-primitives-combobox--basic-usage)

Thanks for spotting that Mark, i added a new test and fixed the issue 😄

joshuaellis commented on packages/primitives/src/components/Combobox/Combobox.tsx:
```diff
@@ -0,0 +1,764 @@
+      disabled={isDisabled}
+      data-disabled={isDisabled ? '' : undefined}
+      data-placeholder={context.textValue === undefined ? '' : undefined}
+      value={context.textValue ?? ''}
```
If you do that you throw the error of trying to control an uncontrolled component :)

joshuaellis commented on packages/primitives/src/components/Combobox/Combobox.tsx:
```diff
@@ -0,0 +1,764 @@
+
+interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {}
+
+const ComboxboxInput = React.forwardRef<HTMLInputElement, InputProps>((props, forwardedRef) => {
```
Sure thing!

joshuaellis commented on packages/primitives/src/components/Combobox/Combobox.tsx:
```diff
@@ -0,0 +1,764 @@
+interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {}
+
+const ComboxboxInput = React.forwardRef<HTMLInputElement, InputProps>((props, forwardedRef) => {
+  const { ...inputProps } = props;
```
That's a hangover.

gu-stav commented:
From functional testing:

- the focus outline of the clear button doesn't look good

<img width="159" alt="Screenshot 2023-02-15 at 11 04 57" src="https://user-images.githubusercontent.com/2244375/218997092-2b7568fb-8125-4c8b-a4b4-82a2eed932e1.png">

- the element has a focus outline even though it is disabled (when clicking on the label)

joshuaellis commented on packages/primitives/src/components/Combobox/Combobox.tsx:
```diff
@@ -0,0 +1,764 @@
+        }
+      })}
+    >
+      {children || '▼'}
```
No I don't think we should, i think this should be as universal as possible. In the DS we'll replace it with our own icon because it's stylised to _our_ needs.

joshuaellis commented on packages/primitives/src/components/Combobox/Combobox.tsx:
```diff
@@ -0,0 +1,764 @@
+  if (!context.open) {
+    const frag = fragment as Element | undefined;
+    return frag
+      ? ReactDOM.createPortal(
+          <Collection.Slot scope={undefined}>
+            <div>{props.children}</div>
+          </Collection.Slot>,
+          frag,
+        )
+      : null;
```
No I don't think we should because we can't add a `fragment` as the mounting point. It would also mean the DS relies on Primitives but Primitives relies on the DS. Its also not used when the menu is open.

gu-stav replied:
True. Wondering if we should move `Portal` to `primitives` too at some point?

joshuaellis replied:
We could do, we'd have to overload the API to allow passing a container as a prop & if one is not supplied, set one up. I'd do this as a separate ticket i think :)

joshuaellis commented:
> From functional testing:
> 
> * the focus outline of the clear button doesn't look good
> 
> <img alt="Screenshot 2023-02-15 at 11 04 57" width="159" src="https://user-images.githubusercontent.com/2244375/218997092-2b7568fb-8125-4c8b-a4b4-82a2eed932e1.png">
> 
> * the element has a focus outline even though it is disabled (when clicking on the label)

I haven't concerted the DS combobox to the Primitive yet, thats another task :)

joshuaellis commented on packages/primitives/src/components/Combobox/Combobox.tsx:
```diff
@@ -0,0 +1,764 @@
+            const items = getItems().filter((item) => !item.disabled);
+            let candidateNodes = items.map((item) => item.ref.current!);
+
+            if (['ArrowUp', 'End'].includes(event.key)) {
+              candidateNodes = candidateNodes.slice().reverse();
+            }
```
yes, because we need to reverse for `ArrowUp`
